community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionHeaders.java

```diff
@@ -0,0 +1,190 @@
+ */
+package org.neo4j.kernel.impl.transaction.log;
+
+import static java.lang.String.format;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.neo4j.kernel.impl.transaction.TransactionRepresentation;
+
+/**
+ * Represents information read from the extra header of a transaction Start entry. It is presented as a map from
+ * integer identifier to byte[], which has a format specific to the handler. The format of the header, all stored in the
+ * byte[] "additionalInformation" field of the Start entry is as follows:
+ * <p>
+ * <pre>
+ * 1 byte for format version
+ * 1 byte for number of entries that follow
+ * for each entry:
+ *     1 byte for identifier
+ *     4 bytes for data length
+ *     byte[] for the header specific data
+ * </pre>
+ * <p>
+ * The above do not include the prefixed 4 bytes that mark the length of the complete array, and which is read by
+ * the Start entry parser. It belongs to a level below this class and is irrelevant to its implementation.
+ *
+ * The creation of the byte array is encapsulated in the {@link TransactionHeadersArrayBuilder}. The parser of the array
+ * is encapsulated in this class. The general usage pattern is that when writing out the header one uses the Builder
+ * to set it up as required and then pass the resulting array to the {@link TransactionAppender} via a
+ * {@link TransactionRepresentation}. Reading back the information requires getting the header array from the
+ * {@link TransactionRepresentation} and feeding it into an instance of this class to get the encoded information.
+ *
+ * The version information kept in the first byte is for future proofing against format changes. Right now there is
+ * only one version, the current one, so it is not used anywhere and defaults always to
+ * TransactionHeaders.CURRENT_VERSION
+ */
+public class TransactionHeaders
+{
+    public static final byte CURRENT_VERSION = 0;
+    private static final byte[] DEFAULT_FOR_NON_EXISTING_ID = new byte[0];
+
+    private final byte version;
+    private final Map<Byte, byte[]> entries = new HashMap<>();
+
+    public TransactionHeaders( byte[] bytes )
+    {
+        this.version = extractVersion( bytes );
+        extractEntries( bytes );
+    }
+
+    /**
+     * @return The number of headers discovered and parsed
+     */
+    public int size()
+    {
+        return entries.size();
+    }
+
+    /**
+     * @return The version this array has been encoded with.
+     */
+    public byte version()
+    {
+        return version;
+    }
+
+    /**
+     * @return The byte array belonging to the specified identifier
+     */
+    public byte[] forIdentifier( byte identifier )
+    {
+        return entries.containsKey( identifier ) ?
+                entries.get( identifier ) :
+                DEFAULT_FOR_NON_EXISTING_ID;
+    }
+
+    private byte extractVersion( byte[] bytes )
+    {
+        return bytes.length > 0 ? bytes[0] : CURRENT_VERSION;
```

Comment:
A comment saying _why_ would be lovely... I'm assuming it's due to backwards compatibility and you and I may know that now, but for future's sake.